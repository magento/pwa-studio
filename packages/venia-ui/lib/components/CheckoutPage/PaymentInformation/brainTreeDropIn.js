/**
 * @fileoverview This component uses the BrainTree Web Drop In to hook into Web
 * Payments and the Payment Request API to submit payments via BrainTree.
 *
 * @see
 *   https://github.com/braintree/braintree-web-drop-in
 *   https://braintree.github.io/braintree-web-drop-in/docs/current/index.html
 *   https://developers.braintreepayments.com/guides/drop-in/overview/javascript/v3
 *   https://developers.braintreepayments.com/guides/payment-method-nonce.
 *   https://braintree.github.io/braintree-web-drop-in/docs/current/Dropin.html#requestPaymentMethod.
 */

import React, { useEffect, useState, useCallback } from 'react';
import { FormattedMessage } from 'react-intl';
import { bool, func, shape, string } from 'prop-types';

import defaultClasses from './braintreeDropin.css';
import { useStyle } from '../../../classify';

const authorization = process.env.CHECKOUT_BRAINTREE_TOKEN;

/**
 * This BraintreeDropin component has two purposes which lend to its
 * implementation:
 *
 * 1) Mount and asynchronously create the dropin via the braintree api.
 * 2) On submission (triggered by a parent), request the payment nonce.
 */
const BraintreeDropin = props => {
    const {
        onError,
        onReady,
        onSuccess,
        shouldRequestPaymentNonce,
        containerId = 'braintree-container',
        shouldTeardownDropin,
        resetShouldTeardownDropin
    } = props;
    const classes = useStyle(defaultClasses, props.classes);
    const [isError, setIsError] = useState(false);
    const [dropinInstance, setDropinInstance] = useState();

    const createDropinInstance = useCallback(async () => {
        const { default: dropIn } = await import('braintree-web-drop-in');
        const dropinInstance = await dropIn.create({
            authorization,
            container: `#${containerId}`,
            card: {
                cardholderName: {
                    required: true
                },
                overrides: {
                    fields: {
                        number: {
                            maskInput: {
                                // Only show last four digits on blur.
                                showLastFour: true
                            }
                        }
                    }
                }
            }
        });

        return dropinInstance;
    }, [containerId]);

    useEffect(() => {
        let unmounted = false;

        const renderDropin = async () => {
            try {
                const instance = await createDropinInstance();

                if (!unmounted) {
                    setDropinInstance(instance);
                    onReady(true);
                } else {
                    /**
                     * Component has been unmounted, tear down the instance.
                     */
                    instance.teardown();
                }
            } catch (err) {
                if (process.env.NODE_ENV !== 'production') {
                    // This error can be common because of the async nature of
                    // the checkout page. If the problem is due to a missing
                    // container it is likely that the component was
                    // intentionally unmounted.
                    console.error(
                        `Unable to initialize Credit Card form (Braintree). \n${err}`
                    );
                }
                if (!unmounted) {
                    setIsError(true);
                }
            }
        };

        renderDropin();

        return () => {
            unmounted = true;
        };
    }, [createDropinInstance, onReady]);

    useEffect(() => {
        async function requestPaymentNonce() {
            try {
                const paymentNonce = await dropinInstance.requestPaymentMethod();
                onSuccess(paymentNonce);
            } catch (e) {
                // An error occurred. BrainTree will update the UI with error
                // messaging, but we should signal that there was an error.
                console.error(`Invalid Payment Details. \n${e}`);
                onError();
            }
        }

        if (shouldRequestPaymentNonce) {
            requestPaymentNonce();
        }
    }, [dropinInstance, onError, onSuccess, shouldRequestPaymentNonce]);

    /**
     * This useEffect handles tearing down and re-creating the dropin
     * in case the parent component needs it to.
     *
     * The parent component does this by setting `shouldTeardownDropin` `true`.
     */
    useEffect(() => {
        let unmounted = false;

        const teardownAndRenderDropin = async () => {
            try {
                dropinInstance.teardown();
                resetShouldTeardownDropin();

                const instance = await createDropinInstance();

                if (!unmounted) {
                    setDropinInstance(instance);
                    onReady(true);
                } else {
                    /**
                     * Component has been unmounted, tear down the instance.
                     */
                    instance.teardown();
                }
            } catch (err) {
                console.error(
                    `Unable to tear down and re-initialize Credit Card form (Braintree). \n${err}`
                );
                if (!unmounted) {
                    setIsError(true);
                }
            }
        };

        if (shouldTeardownDropin) {
            teardownAndRenderDropin();
        }

        return () => {
            unmounted = true;
        };
    }, [
        shouldTeardownDropin,
        dropinInstance,
        resetShouldTeardownDropin,
        createDropinInstance,
        onReady
    ]);

    if (isError) {
        return (
            <span className={classes.error}>
                <FormattedMessage
                    id={'checkoutPage.errorLoadingPayment'}
                    defaultMessage={
                        'There was an error loading payment options. Please try again later.'
                    }
                />
            </span>
        );
    }

    return (
        <div className={classes.root}>
            <div id={containerId} />
        </div>
    );
};

export default BraintreeDropin;

BraintreeDropin.propTypes = {
    classes: shape({
        root: string,
        error: string
    }),
    containerId: string,
    onError: func.isRequired,
    onReady: func.isRequired,
    onSuccess: func.isRequired,
    shouldRequestPaymentNonce: bool.isRequired
};
