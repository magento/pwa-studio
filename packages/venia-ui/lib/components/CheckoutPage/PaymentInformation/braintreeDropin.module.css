.root {
}

.error {
    color: var(--venia-global-color-error);
}

div[data-braintree-id='card'] {
    border: none;
}

div[data-braintree-id='card-sheet-header'] {
    display: none;
}

div[class*='braintree-sheet__content--form'] {
    /**
      This is needed to override padding on the dropin
      component. If we use padding-top instead it just
      overrides the padding-top not whole padding.
    */
    padding: 1rem 0rem 0rem 0rem;
}
