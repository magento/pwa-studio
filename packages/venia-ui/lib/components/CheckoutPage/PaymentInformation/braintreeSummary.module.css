.root {
    display: grid;
    gap: 1.125rem;
    padding: 2rem;
}

.loading {
    composes: root from '../../LoadingIndicator/indicator.module.css';
    text-align: center;
}

.heading_container {
    display: grid;
    grid-auto-flow: column;
    grid-template-columns: 1fr;
}

.edit_button {
    color: rgb(var(--venia-brand-color-1-700));
    padding: 1rem;
    margin: -1rem;
}

.heading {
    font-weight: 600;
}

.address_summary_container {
    display: grid;
    row-gap: 0.5rem;
}

.first_name {
    padding-right: 5px;
}

.last_name {
    padding-right: 5px;
}

.street1 {
    padding-right: 5px;
}

.street2 {
    padding-right: 5px;
}

.city {
    padding-right: 5px;
}

.state {
    padding-right: 5px;
}

.postalCode {
    padding-right: 5px;
}

.country {
    padding-right: 5px;
}

.edit_icon {
    stroke: rgb(var(--venia-brand-color-1-700));
}

.card_details_container {
    display: grid;
    gap: 0.5rem;
}

.payment_type {
}

.payment_details {
}

@media (max-width: 960px) {
    .edit_text {
        display: none;
    }
}
