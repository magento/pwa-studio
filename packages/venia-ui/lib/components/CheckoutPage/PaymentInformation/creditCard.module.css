.root {
}

.credit_card_root {
    visibility: visible;
    opacity: 1;
    transition-delay: 64ms;
    transition-duration: 384ms;
    transition-property: opacity, visbility;
    transition-timing-function: var(--venia-global-anim-standard);
}

.credit_card_root_hidden {
    overflow: hidden;
    visibility: hidden;
    height: 0;
    opacity: 0;
}

.dropin_root {
}

.billing_address_fields_root {
    display: grid;
    gap: 0.9375rem 1.5rem;
    padding: 1rem 0rem;
}

.billing_address_fields_root_hidden {
    overflow: hidden;
    visibility: hidden;
    height: 0;
    opacity: 0;
}

.formErrorContainer {
    padding-top: 1.5rem;
}

.first_name,
.last_name {
    grid-column-end: span 1;
}

.country,
.street1,
.street2,
.address_check,
.city,
.region,
.postal_code,
.phone_number {
    grid-column-end: span 2;
}

/*
 * Mobile-specific styles.
 */

@media (max-width: 960px) {
    .first_name,
    .last_name,
    .city,
    .region,
    .postal_code,
    .phone_number,
    .street1,
    .street2,
    .address_check {
        grid-column-end: span 2;
    }
}
