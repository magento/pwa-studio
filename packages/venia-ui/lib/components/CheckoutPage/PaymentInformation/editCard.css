.root {
    overflow: auto;
    padding: 0.5rem 1rem;
}
