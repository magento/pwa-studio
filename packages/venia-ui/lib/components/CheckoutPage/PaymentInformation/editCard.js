import React from 'react';
import { shape, string, bool, func } from 'prop-types';
import defaultClasses from './editCard.module.css';
import { useStyle } from '../../../classify';
import CreditCard from './creditCard';

/**
 * The edit view for the Braintree payment method.
 */
const EditCard = props => {
    const classes = useStyle(defaultClasses, props.classes);
    const {
        onPaymentReady,
        onPaymentSuccess,
        onPaymentError,
        resetShouldSubmit,
        shouldSubmit
    } = props;

    return (
        <div className={classes.root}>
            <CreditCard
                onPaymentReady={onPaymentReady}
                onPaymentSuccess={onPaymentSuccess}
                onPaymentError={onPaymentError}
                resetShouldSubmit={resetShouldSubmit}
                shouldSubmit={shouldSubmit}
            />
        </div>
    );
};

EditCard.propTypes = {
    classes: shape({
        root: string
    }),
    onPaymentReady: func.isRequired,
    onPaymentSuccess: func.isRequired,
    onPaymentError: func.isRequired,
    resetShouldSubmit: func.isRequired,
    shouldSubmit: bool
};

export default EditCard;
