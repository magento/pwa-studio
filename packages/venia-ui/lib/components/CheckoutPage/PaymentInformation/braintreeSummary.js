import React from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { useStyle } from '../../../classify';
import { shape, string, func, bool } from 'prop-types';
import { Edit2 as EditIcon } from 'react-feather';
import Icon from '../../Icon';
import LinkButton from '../../LinkButton';

import { useBraintreeSummary } from '@magento/peregrine/lib/talons/CheckoutPage/PaymentInformation/useBraintreeSummary';
import defaultClasses from './braintreeSummary.css';

const BraintreeSummary = props => {
    const { classes: propClasses, onEdit } = props;

    const classes = useStyle(defaultClasses, propClasses);
    const { formatMessage } = useIntl();

    const {
        billingAddress,
        isBillingAddressSame,
        isLoading,
        paymentNonce
    } = useBraintreeSummary();

    if (isLoading) {
        return (
            <LoadingIndicator classes={{ root: classes.loading }}>
                <FormattedMessage
                    id={'checkoutPage.loadingPaymentInformation'}
                    defaultMessage={'Fetching Payment Information'}
                />
            </LoadingIndicator>
        );
    }

    const paymentSummary = formatMessage(
        {
            id: 'checkoutPage.paymentSummary',
            defaultMessage: 'Card'
        },
        {
            cardType: paymentNonce.details.cardType,
            lastFour: paymentNonce.details.lastFour
        }
    );

    const billingAddressSummary =
        !isBillingAddressSame && billingAddress ? (
            <div className={classes.address_summary_container}>
                <div>
                    <span className={classes.first_name}>
                        {billingAddress.firstName}
                    </span>
                    <span className={classes.last_name}>
                        {billingAddress.lastName}
                    </span>
                </div>
                <div>
                    <span className={classes.street1}>
                        {billingAddress.street1}
                    </span>
                    <span className={classes.street2}>
                        {billingAddress.street2}
                    </span>
                    <span className={classes.city}>{billingAddress.city}</span>
                    <span className={classes.state}>
                        {billingAddress.state}
                    </span>
                </div>
                <div>
                    <span className={classes.postalCode}>
                        {billingAddress.postalCode}
                    </span>
                    <span className={classes.country}>
                        {billingAddress.country}
                    </span>
                </div>
            </div>
        ) : null;

    return (
        <div className={classes.root}>
            <div className={classes.heading_container}>
                <h5 className={classes.heading}>
                    <FormattedMessage
                        id={'checkoutPage.paymentInformation'}
                        defaultMessage={'Payment Information'}
                    />
                </h5>
                <LinkButton
                    className={classes.edit_button}
                    onClick={onEdit}
                    type="button"
                >
                    <Icon
                        size={16}
                        src={EditIcon}
                        classes={{ icon: classes.edit_icon }}
                    />
                    <span className={classes.edit_text}>
                        <FormattedMessage
                            id={'global.editButton'}
                            defaultMessage={'Edit'}
                        />
                    </span>
                </LinkButton>
            </div>
            <div className={classes.card_details_container}>
                <span className={classes.payment_type}>
                    <FormattedMessage
                        id={'global.creditCard'}
                        defaultMessage={'Credit Card'}
                    />
                </span>
                <span className={classes.payment_details}>
                    {paymentSummary}
                </span>
            </div>
            {billingAddressSummary}
        </div>
    );
};

export default BraintreeSummary;

BraintreeSummary.propTypes = {
    classes: shape({
        root: string,
        heading_container: string,
        heading: string,
        edit_button: string,
        card_details_container: string,
        address_summary_container: string,
        first_name: string,
        last_name: string,
        street1: string,
        street2: string,
        city: string,
        postalCode: string,
        country: string,
        payment_type: string,
        payment_details: string
    }),
    billingAddress: shape({
        firstName: string,
        lastName: string,
        country: string,
        street1: string,
        street2: string,
        city: string,
        state: string,
        postalCode: string
    }),
    onEdit: func.isRequired,
    isBillingAddressSame: bool
};
