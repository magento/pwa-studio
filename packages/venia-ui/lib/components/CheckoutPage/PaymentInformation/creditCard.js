import React, { useMemo, useCallback } from 'react';
import { useIntl } from 'react-intl';
import { bool, func, shape, string } from 'prop-types';
import { useCreditCard } from '@magento/peregrine/lib/talons/CheckoutPage/PaymentInformation/useCreditCard';

import { isRequired } from '../../../util/formValidators';
import Country from '../../Country';
import Region from '../../Region';
import Postcode from '../../Postcode';
import Checkbox from '../../Checkbox';
import Field from '../../Field';
import TextInput from '../../TextInput';
import BrainTreeDropin from './brainTreeDropIn';
import LoadingIndicator from '../../LoadingIndicator';
import { useStyle } from '../../../classify';

import defaultClasses from './creditCard.module.css';
import FormError from '../../FormError';

const STEP_DESCRIPTIONS = [
    { defaultMessage: 'Loading Payment', id: 'checkoutPage.step0' },
    {
        defaultMessage: 'Checking Credit Card Information',
        id: 'checkoutPage.step1'
    },
    {
        defaultMessage: 'Checking Credit Card Information',
        id: 'checkoutPage.step2'
    },
    {
        defaultMessage: 'Checking Credit Card Information',
        id: 'checkoutPage.step3'
    },
    {
        defaultMessage: 'Saved Credit Card Information Successfully',
        id: 'checkoutPage.step4'
    }
];

/**
 * The initial view for the Braintree payment method.
 */
const CreditCard = props => {
    const {
        classes: propClasses,
        onPaymentSuccess: onSuccess,
        onPaymentReady: onReady,
        onPaymentError: onError,
        resetShouldSubmit,
        shouldSubmit
    } = props;
    const { formatMessage } = useIntl();

    const classes = useStyle(defaultClasses, propClasses);

    const talonProps = useCreditCard({
        onSuccess,
        onReady,
        onError,
        shouldSubmit,
        resetShouldSubmit
    });

    const {
        errors,
        shouldRequestPaymentNonce,
        onPaymentError,
        onPaymentSuccess,
        onPaymentReady,
        isBillingAddressSame,
        isLoading,
        /**
         * `stepNumber` depicts the state of the process flow in credit card
         * payment flow.
         *
         * `0` No call made yet
         * `1` Billing address mutation intiated
         * `2` Braintree nonce requsted
         * `3` Payment information mutation intiated
         * `4` All mutations done
         */
        stepNumber,
        initialValues,
        shippingAddressCountry,
        shouldTeardownDropin,
        resetShouldTeardownDropin
    } = talonProps;

    const creditCardComponentClassName = isLoading
        ? classes.credit_card_root_hidden
        : classes.credit_card_root;

    const billingAddressFieldsClassName = isBillingAddressSame
        ? classes.billing_address_fields_root_hidden
        : classes.billing_address_fields_root;

    /**
     * Instead of defining classes={root: classes.FIELD_NAME}
     * we are using useMemo to only do it once (hopefully).
     */
    const fieldClasses = useMemo(() => {
        return [
            'first_name',
            'last_name',
            'country',
            'street1',
            'street2',
            'city',
            'region',
            'postal_code',
            'phone_number'
        ].reduce((acc, fieldName) => {
            acc[fieldName] = { root: classes[fieldName] };

            return acc;
        }, {});
    }, [classes]);

    /**
     * These 2 functions are wrappers around the `isRequired` function
     * of `formValidators`. They perform validations only if the
     * billing address is different from shipping address.
     *
     * We write this function in `venia-ui` and not in the `peregrine` talon
     * because it references `isRequired` which is a `venia-ui` function.
     */
    const isFieldRequired = useCallback((value, { isBillingAddressSame }) => {
        if (isBillingAddressSame) {
            /**
             * Informed validator functions return `undefined` if
             * validation is `true`
             */
            return undefined;
        } else {
            return isRequired(value);
        }
    }, []);

    const stepTitle = STEP_DESCRIPTIONS[stepNumber].defaultMessage
        ? formatMessage({
              id: STEP_DESCRIPTIONS[stepNumber].id,
              defaultMessage: STEP_DESCRIPTIONS[stepNumber].defaultMessage
          })
        : formatMessage({
              id: 'checkoutPage.loadingPayment',
              defaultMessage: 'Loading Payment'
          });

    const loadingIndicator = isLoading ? (
        <LoadingIndicator>{stepTitle}</LoadingIndicator>
    ) : null;

    return (
        <div className={classes.root} data-cy="CreditCard-root">
            <div className={creditCardComponentClassName}>
                <FormError
                    classes={{ root: classes.formErrorContainer }}
                    errors={Array.from(errors.values())}
                />
                <div className={classes.dropin_root}>
                    <BrainTreeDropin
                        onError={onPaymentError}
                        onReady={onPaymentReady}
                        onSuccess={onPaymentSuccess}
                        shouldRequestPaymentNonce={shouldRequestPaymentNonce}
                        shouldTeardownDropin={shouldTeardownDropin}
                        resetShouldTeardownDropin={resetShouldTeardownDropin}
                    />
                </div>
                <div className={classes.address_check}>
                    <Checkbox
                        data-cy="PaymentInformation-billingAddressSame"
                        field="isBillingAddressSame"
                        label={formatMessage({
                            id: 'checkoutPage.billingAddressSame',
                            defaultMessage:
                                'Billing address same as shipping address'
                        })}
                        initialValue={initialValues.isBillingAddressSame}
                    />
                </div>
                <div className={billingAddressFieldsClassName}>
                    <Field
                        id="firstName"
                        classes={fieldClasses.first_name}
                        label={formatMessage({
                            id: 'global.firstName',
                            defaultMessage: 'First Name'
                        })}
                    >
                        <TextInput
                            data-cy="CreditCard-billingAddress-firstname"
                            id="firstName"
                            field="firstName"
                            validate={isFieldRequired}
                            initialValue={initialValues.firstName}
                        />
                    </Field>
                    <Field
                        id="lastName"
                        classes={fieldClasses.last_name}
                        label={formatMessage({
                            id: 'global.lastName',
                            defaultMessage: 'Last Name'
                        })}
                    >
                        <TextInput
                            data-cy="CreditCard-billingAddress-lastname"
                            id="lastName"
                            field="lastName"
                            validate={isFieldRequired}
                            initialValue={initialValues.lastName}
                        />
                    </Field>
                    <Country
                        data-cy="CreditCard-billingAddress-country"
                        classes={fieldClasses.country}
                        validate={isFieldRequired}
                        initialValue={
                            /**
                             * If there is no initial value to start with
                             * use the country from shipping address.
                             */
                            initialValues.country || shippingAddressCountry
                        }
                    />
                    <Field
                        id="street1"
                        classes={fieldClasses.street1}
                        label={formatMessage({
                            id: 'global.streetAddress',
                            defaultMessage: 'Street Address'
                        })}
                    >
                        <TextInput
                            data-cy="CreditCard-billingAddress-street1"
                            id="street1"
                            field="street1"
                            validate={isFieldRequired}
                            initialValue={initialValues.street1}
                        />
                    </Field>
                    <Field
                        id="street2"
                        classes={fieldClasses.street2}
                        label={formatMessage({
                            id: 'global.streetAddress2',
                            defaultMessage: 'Street Address 2'
                        })}
                        optional={true}
                    >
                        <TextInput
                            data-cy="CreditCard-billingAddress-street2"
                            id="street2"
                            field="street2"
                            initialValue={initialValues.street2}
                        />
                    </Field>
                    <Field
                        id="city"
                        classes={fieldClasses.city}
                        label={formatMessage({
                            id: 'global.city',
                            defaultMessage: 'City'
                        })}
                    >
                        <TextInput
                            data-cy="CreditCard-billingAddress-city"
                            id="city"
                            field="city"
                            validate={isFieldRequired}
                            initialValue={initialValues.city}
                        />
                    </Field>
                    <Region
                        data-cy="CreditCard-billingAddress-region"
                        classes={fieldClasses.region}
                        initialValue={initialValues.region}
                        validate={isFieldRequired}
                        fieldInput={'region[label]'}
                        fieldSelect={'region[region_id]'}
                        optionValueKey={'id'}
                    />
                    <Postcode
                        data-cy="CreditCard-billingAddress-postcode"
                        classes={fieldClasses.postal_code}
                        validate={isFieldRequired}
                        initialValue={initialValues.postcode}
                    />
                    <Field
                        id="phoneNumber"
                        classes={fieldClasses.phone_number}
                        label={formatMessage({
                            id: 'global.phoneNumber',
                            defaultMessage: 'Phone Number'
                        })}
                    >
                        <TextInput
                            data-cy="CreditCard-billingAddress-phoneNumber"
                            id="phoneNumber"
                            field="phoneNumber"
                            validate={isFieldRequired}
                            initialValue={initialValues.phoneNumber}
                        />
                    </Field>
                </div>
            </div>
            {loadingIndicator}
        </div>
    );
};

export default CreditCard;

CreditCard.propTypes = {
    classes: shape({
        root: string,
        dropin_root: string,
        billing_address_fields_root: string,
        first_name: string,
        last_name: string,
        city: string,
        region: string,
        postal_code: string,
        phone_number: string,
        country: string,
        street1: string,
        street2: string,
        address_check: string,
        credit_card_root: string,
        credit_card_root_hidden: string
    }),
    shouldSubmit: bool.isRequired,
    onPaymentSuccess: func,
    onPaymentReady: func,
    onPaymentError: func,
    resetShouldSubmit: func.isRequired
};
