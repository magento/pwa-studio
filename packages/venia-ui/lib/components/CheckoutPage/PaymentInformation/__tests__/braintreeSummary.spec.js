import React from 'react';
import createTestInstance from '@magento/peregrine/lib/util/createTestInstance';
jest.mock('../../../../classify');

import classes from '../braintreeSummary.css';
import { useBraintreeSummary } from '@magento/peregrine/lib/talons/CheckoutPage/PaymentInformation/useBraintreeSummary';
import BraintreeSummary from '../braintreeSummary';

const billingAddress = {
    firstName: 'Goosey',
    lastName: 'Goose',
    country: 'Goose Land',
    street1: '12345 Goosey Blvd',
    street2: 'Apt 123',
    city: 'Austin',
    state: 'Texas',
    postalCode: '12345',
    phoneNumber: '1234657890'
};

const paymentNonce = {
    details: { cardType: 'visa', lastFour: '1234' }
};

jest.mock(
    '@magento/peregrine/lib/talons/CheckoutPage/PaymentInformation/useBraintreeSummary',
    () => {
        return {
            useBraintreeSummary: jest.fn(() => ({
                billingAddress: {
                    firstName: 'Goosey',
                    lastName: 'Goose',
                    country: 'Goose Land',
                    street1: '12345 Goosey Blvd',
                    street2: 'Apt 123',
                    city: 'Austin',
                    state: 'Texas',
                    postalCode: '12345',
                    phoneNumber: '1234657890'
                },
                isBillingAddressSame: true,
                isLoading: false,
                paymentNonce: {
                    details: { cardType: 'visa', lastFour: '1234' }
                }
            }))
        };
    }
);

test('should render', () => {
    useBraintreeSummary.mockReturnValueOnce({
        billingAddress,
        isBillingAddressSame: true,
        isLoading: false,
        paymentNonce
    });
    const tree = createTestInstance(<BraintreeSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
});

test('Should render billing address if it is not same as shipping address', () => {
    useBraintreeSummary.mockReturnValueOnce({
        billingAddress,
        isBillingAddressSame: false,
        isLoading: false,
        paymentNonce
    });
    const tree = createTestInstance(<BraintreeSummary />);

    expect(tree.toJSON()).toMatchSnapshot();

    expect(
        tree.root.findByProps({
            className: classes.address_summary_container
        })
    ).not.toBeNull();
});

test('Should not render billing address if it is same as shipping address', () => {
    useBraintreeSummary.mockReturnValueOnce({
        billingAddress: {},
        isBillingAddressSame: false,
        isLoading: false,
        paymentNonce
    });
    const tree = createTestInstance(<BraintreeSummary />);

    expect(tree.toJSON()).toMatchSnapshot();
    expect(
        tree.root.findByProps({
            className: classes.address_summary_container
        })
    ).not.toBeNull();
});
