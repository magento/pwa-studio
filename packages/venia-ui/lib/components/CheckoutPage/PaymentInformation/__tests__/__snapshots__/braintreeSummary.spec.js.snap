// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should not render billing address if it is same as shipping address 1`] = `
<div
  className="root"
>
  <div
    className="heading_container"
  >
    <h5
      className="heading"
    >
      <mock-FormattedMessage
        defaultMessage="Payment Information"
        id="checkoutPage.paymentInformation"
      />
    </h5>
    <button
      className="edit_button"
      disabled={false}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <span
        className="content"
      >
        <span
          className="root"
        >
          <svg
            className="icon"
            fill="none"
            height={16}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
            />
          </svg>
        </span>
        <span
          className="edit_text"
        >
          <mock-FormattedMessage
            defaultMessage="Edit"
            id="global.editButton"
          />
        </span>
      </span>
    </button>
  </div>
  <div
    className="card_details_container"
  >
    <span
      className="payment_type"
    >
      <mock-FormattedMessage
        defaultMessage="Credit Card"
        id="global.creditCard"
      />
    </span>
    <span
      className="payment_details"
    >
      visa ending in 1234
    </span>
  </div>
  <div
    className="address_summary_container"
  >
    <div>
      <span
        className="first_name"
      />
      <span
        className="last_name"
      />
    </div>
    <div>
      <span
        className="street1"
      />
      <span
        className="street2"
      />
      <span
        className="city"
      />
      <span
        className="state"
      />
    </div>
    <div>
      <span
        className="postalCode"
      />
      <span
        className="country"
      />
    </div>
  </div>
</div>
`;

exports[`Should render billing address if it is not same as shipping address 1`] = `
<div
  className="root"
>
  <div
    className="heading_container"
  >
    <h5
      className="heading"
    >
      <mock-FormattedMessage
        defaultMessage="Payment Information"
        id="checkoutPage.paymentInformation"
      />
    </h5>
    <button
      className="edit_button"
      disabled={false}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <span
        className="content"
      >
        <span
          className="root"
        >
          <svg
            className="icon"
            fill="none"
            height={16}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
            />
          </svg>
        </span>
        <span
          className="edit_text"
        >
          <mock-FormattedMessage
            defaultMessage="Edit"
            id="global.editButton"
          />
        </span>
      </span>
    </button>
  </div>
  <div
    className="card_details_container"
  >
    <span
      className="payment_type"
    >
      <mock-FormattedMessage
        defaultMessage="Credit Card"
        id="global.creditCard"
      />
    </span>
    <span
      className="payment_details"
    >
      visa ending in 1234
    </span>
  </div>
  <div
    className="address_summary_container"
  >
    <div>
      <span
        className="first_name"
      >
        Goosey
      </span>
      <span
        className="last_name"
      >
        Goose
      </span>
    </div>
    <div>
      <span
        className="street1"
      >
        12345 Goosey Blvd
      </span>
      <span
        className="street2"
      >
        Apt 123
      </span>
      <span
        className="city"
      >
        Austin
      </span>
      <span
        className="state"
      >
        Texas
      </span>
    </div>
    <div>
      <span
        className="postalCode"
      >
        12345
      </span>
      <span
        className="country"
      >
        Goose Land
      </span>
    </div>
  </div>
</div>
`;

exports[`should render 1`] = `
<div
  className="root"
>
  <div
    className="heading_container"
  >
    <h5
      className="heading"
    >
      <mock-FormattedMessage
        defaultMessage="Payment Information"
        id="checkoutPage.paymentInformation"
      />
    </h5>
    <button
      className="edit_button"
      disabled={false}
      onDragStart={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <span
        className="content"
      >
        <span
          className="root"
        >
          <svg
            className="icon"
            fill="none"
            height={16}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17 3a2.828 2.828 0 1 1 4 4L7.5 20.5 2 22l1.5-5.5L17 3z"
            />
          </svg>
        </span>
        <span
          className="edit_text"
        >
          <mock-FormattedMessage
            defaultMessage="Edit"
            id="global.editButton"
          />
        </span>
      </span>
    </button>
  </div>
  <div
    className="card_details_container"
  >
    <span
      className="payment_type"
    >
      <mock-FormattedMessage
        defaultMessage="Credit Card"
        id="global.creditCard"
      />
    </span>
    <span
      className="payment_details"
    >
      visa ending in 1234
    </span>
  </div>
</div>
`;
